optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/domain/rotation/TimeBucket.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaweb.employeerostering.domain.rotation;
+
+import java.time.DayOfWeek;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.OffsetDateTime;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.persistence.CollectionTable;
+import javax.persistence.ElementCollection;
+import javax.persistence.Entity;
+import javax.persistence.ForeignKey;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.OneToMany;
+
+import org.optaweb.employeerostering.domain.common.AbstractPersistable;
+import org.optaweb.employeerostering.domain.shift.Shift;
+import org.optaweb.employeerostering.domain.skill.Skill;
+import org.optaweb.employeerostering.domain.spot.Spot;
+
+@Entity
+public class TimeBucket extends AbstractPersistable {
+    private LocalTime startTime;
+    private LocalTime endTime;
+    
+    @ManyToOne
+    private Spot spot;
+    
+    @OneToMany
+    private Set<Skill> additionalSkillSet;
+    
+    @ElementCollection
+    @CollectionTable(name = "repeat_on_day_set",
+        joinColumns = {
+        @JoinColumn(name = "day_id",
+            referencedColumnName = "id",
+            foreignKey=@ForeignKey(name="DAY_FK",
+            foreignKeyDefinition = "FOREIGN KEY (day_id) references public.time_bucket (id)" +
+            " ON UPDATE NO ACTION ON DELETE CASCADE"))
+        }
+    )
+    private Set<DayOfWeek> repeatOnDaySet;
+    
+    @ElementCollection
+    @CollectionTable(name = "seat_list",
+        joinColumns = {
+        @JoinColumn(name = "seat_id",
+            referencedColumnName = "id",
+            foreignKey=@ForeignKey(name="SEAT_FK",
+            foreignKeyDefinition = "FOREIGN KEY (seat_id) references public.time_bucket (id)" +
+            " ON UPDATE NO ACTION ON DELETE CASCADE"))
+        }
+    )
+    private List<Seat> seatList;
```

Comment:
I am not familiar with this part of JPA, but the syntax doesn't look particularly DB-agnostic. Which DBs did you check?

Reply:
Hibernate/H2. Getting Cascade delete to work on an ElementCollection is non-trivial (see https://stackoverflow.com/questions/7695831/how-can-i-cascade-delete-a-collection-which-is-part-of-a-jpa-entity). I might be able to change it to https://stackoverflow.com/a/62848296

Reply:
Sadly, that answer use hibernate specific annotation, so I am keeping it as is.